planner/core/rule_join_reorder_greedy.go

```diff
@@ -106,31 +173,46 @@ func (s *joinReorderGreedySolver) constructConnectedJoinTree(tracer *joinReorder
 		}
 		s.curJoinGroup = append(s.curJoinGroup[:bestIdx], s.curJoinGroup[bestIdx+1:]...)
 		s.otherConds = finalRemainOthers
+		s.remainJoinGroup = append(s.remainJoinGroup[0:bestIdx], s.remainJoinGroup[bestIdx+1:]...)
 	}
 	return curJoinTree, nil
 }
 
 func (s *joinReorderGreedySolver) checkConnectionAndMakeJoin(leftNode, rightNode LogicalPlan) (LogicalPlan, []expression.Expression) {
 	var usedEdges []*expression.ScalarFunction
+
```

Comment:
We should check the connection here, for the case that A left join B on A.xx = B.xx left join C on B.xx=C.xx and C.xx=A.xx. And the sort result is [A, B, C] and the only order is {{A, B}, C}.

But if {A, C}'s result is smaller than {A, B} we still might choose {A, C} if we don't check the connection status.